construct-sturdy-uis-with-xstate/notes/04-block-a-state-transition-with-a-guard.md

```diff
@@ -0,0 +1,77 @@
+# Block a State Transition with a Guard
+
+- [Video Link](https://egghead.io/lessons/react-block-a-state-transition-with-a-guard)
+- [Code Link](https://github.com/isaacplmann/sturdy-uis/tree/lesson3-end)
+
+What do we do when we want state transitions to only occur when a certain condition is met?
+
+- We can add [guards](https://xstate.js.org/docs/guides/guards.html#guards-condition-functions)
+
+A good use case to add guards is when we can be faced with two new states.
+
+- If the data structure returns with data inside
+- If the data structure returns empty
+
+📝Guards can and should be [serialized](https://github.com/isaacplmann/sturdy-uis/tree/lesson5-end) as a string or object with the `{ type: '' }` property
+  - The guard logic can also be serialized in a [custom guard](https://xstate.js.org/docs/guides/guards.html#custom-guards).
+
+First, we need to tell `xstate` what are the possible states:
+
+```js
+// informs state of the possible successful states
+successful: {
+    on: {
+        FETCH: 'pending' 
+    },
+    states: {
+        withData: {}, 👈
+        withoutData: {} 👈
+    }
+}
+```
+
+* On `successful` response, from `pending` state, `xstate` will decide to which state to move onto based on if the response has `data` or not.
+
+We can achieve this by adding an array of the possible outcomes to our `pending` state, as such:
+
+```js
+pending: {
+	invoke: {
+		src: 'fetchData',              👇
+        onDone: [{ target: 'successful.withData', actions: ['setResults'], cond:
+        👉 'hasData' }. { target: 'successful.withoutData', actions: ['setResults'] }], 
+		onError: [{ target: 'failed', actions: ['setMessage'] }]
+	}
+},
+```
+
+A [cond](https://xstate.js.org/docs/guides/guards.html#guards-condition-functions) function returns a boolean which determines whether the transition should be allowed to take place.
+  - When the condition is met it will proceed to the corresponding state, in this case the `cond` would be `hasData` and if the response in has data the condition would be met.
```

Comment:
```suggestion
  - When the condition is met it will proceed to the corresponding state, in this case the `cond` would be `hasData` and if the response has data the condition would be met.
```